src/main/scala/vsys/blockchain/contract/DataType.scala

```diff
@@ -1,34 +1,162 @@
-  val Balance         = DataTypeVal(12, true, AmountLength)
-  val OpcBlock        = DataTypeVal(13, false, MaxOpcBlockLength)
-
-  def fromByte(b: Byte): Option[DataType.DataTypeVal] = Try(DataType(b).asInstanceOf[DataTypeVal]).toOption
+  val ContractAccount = DataTypeVal[vsys.account.ContractAccount](6,
+    lenFixed          = true,
+    maxLen            = vsys.account.ContractAccount.AddressLength,
+    deserializer      = vsys.account.ContractAccount.fromBytes(_).right.get,
+    serializer        = a => a.bytes.arr,
+    jsonifier         = a => Json.toJson(a.address),
+    extValidator      = vsys.account.ContractAccount.fromBytes(_).isRight)
+
+  val Account         = DataTypeVal[vsys.account.Account](7,
+    lenFixed          = false,
+    maxLen            = (vsys.account.Address.AddressLength).max(vsys.account.ContractAccount.AddressLength),
+    deserializer      = vsys.account.Account.fromBytes(_, 0).right.get._1,
+    serializer        = a => a.bytes.arr,
+    jsonifier         = a => Json.toJson(a.bytes.arr),
+    extValidator      = vsys.account.Account.fromBytes(_, 0).isRight)
+
+  val TokenId         = DataTypeVal[ByteStr](8,
+    lenFixed          = true,
+    maxLen            = vsys.account.ContractAccount.TokenAddressLength,
+    deserializer      = ByteStr(_),
+    serializer        = b => b.arr,
+    jsonifier         = b => Json.toJson(b.base58),
+    extValidator      = addressBytes => {
+      // length is already ensured here
+      def scheme    = AddressScheme.current
+      val version   = addressBytes.head
+      val network   = addressBytes.tail.head
+      val checkSum1 = addressBytes.takeRight(ChecksumLength)
+      val checkSum2 = hash((addressBytes.dropRight(ChecksumLength))).take(ChecksumLength)
+      // addr version && net id && checksum
+      version == vsys.account.ContractAccount.TokenAddressVersion && network == scheme.chainId && checkSum1.sameElements(checkSum2)
+    })
+
+  val Timestamp       = DataTypeVal[Long](9,
+    lenFixed          = true,
+    maxLen            = TimestampLength,
+    deserializer      = Longs.fromByteArray(_),
+    serializer        = Longs.toByteArray(_),
+    jsonifier         = Json.toJson(_),
+    extValidator      = Longs.fromByteArray(_) >= 0)
+
+  val Boolean         = DataTypeVal[Boolean](10,
+    lenFixed          = true,
+    maxLen            = 1,
+    deserializer      = b => b.head == 1,
+    serializer        = b => Array((if(b) 1 else 0).toByte),
+    jsonifier         = b => Json.toJson(b.toString),
+    extValidator      = b => (b(0) == 1 || b(0) == 0))
+
+  val ShortBytes      = DataTypeVal[Array[Byte]](11,
+    lenFixed          = false,
+    maxLen            = MaxShortBytesLength,
+    deserializer      = b => b.drop(2),
+    serializer        = b => arrayShortLengthToByteArray(b) ++ b,
+    jsonifier         = b => Json.toJson(Base58.encode(b)),
+    extValidator      = _ => true)
+
+  val Balance         = DataTypeVal[Long](12,
```

Comment:
I think we do not use `Balance` at the current stage, may remove it first. Or put it at last with `extValidator = _ => false`.

Reply:
OK